cppcache/src/PdxHelper.cpp

```diff
@@ -47,92 +47,105 @@ PdxHelper::PdxHelper() {}
 
 PdxHelper::~PdxHelper() {}
 
-void PdxHelper::serializePdx(
-    DataOutput& output, const std::shared_ptr<PdxSerializable>& pdxObject) {
-  auto pdxII = std::dynamic_pointer_cast<PdxInstanceImpl>(pdxObject);
-  auto cacheImpl = CacheRegionHelper::getCacheImpl(output.getCache());
-  auto pdxTypeRegistry = cacheImpl->getPdxTypeRegistry();
-  auto& cachePerfStats = cacheImpl->getCachePerfStats();
+std::shared_ptr<PdxInstanceImpl> PdxHelper::getPdxInstance(
+    const std::shared_ptr<PdxSerializable>& pdxObject) {
+  return std::dynamic_pointer_cast<PdxInstanceImpl>(pdxObject);
+}
 
-  if (pdxII != nullptr) {
-    auto piPt = pdxII->getPdxType();
-    if (piPt != nullptr &&
-        piPt->getTypeId() ==
-            0)  // from pdxInstance factory need to get typeid from server
-    {
-      int typeId = pdxTypeRegistry->getPDXIdForType(
-          piPt, DataOutputInternal::getPool(output));
-      pdxII->setPdxId(typeId);
-    }
-    auto plw = PdxLocalWriter(output, piPt, pdxTypeRegistry);
-    pdxII->toData(plw);
-    plw.endObjectWriting();  // now write typeid
-    int len = 0;
-    uint8_t* pdxStream = plw.getPdxStream(len);
-    pdxII->updatePdxStream(pdxStream, len);
+bool PdxHelper::isPdxInstance(
+    const std::shared_ptr<PdxSerializable>& pdxObject) {
+  return (getPdxInstance(pdxObject) != nullptr);
+}
 
-    delete[] pdxStream;
+bool PdxHelper::hasPdxTypeId(
+    const std::shared_ptr<PdxInstanceImpl>& pdxObject) {
+  auto objectType = pdxObject->getPdxType();
+  return (objectType && (objectType->getTypeId() != 0));
+}
 
-    return;
+void PdxHelper::serializePdxInstance(
+    const std::shared_ptr<PdxInstanceImpl>& pdxInstance,
+    std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry, DataOutput& output) {
+  if (!hasPdxTypeId(pdxInstance)) {
+    int typeId = pdxTypeRegistry->getPDXIdForType(
+        pdxInstance->getPdxType(), DataOutputInternal::getPool(output));
+    pdxInstance->setPdxId(typeId);
   }
+  auto pdxLocalWriter =
+      PdxLocalWriter(output, pdxInstance->getPdxType(), pdxTypeRegistry);
+  pdxInstance->toData(pdxLocalWriter);
+  pdxLocalWriter.endObjectWriting();  // now write typeid
+  auto pdxStream = pdxLocalWriter.getPdxStream();
+  pdxInstance->updatePdxStream(pdxStream, pdxLocalWriter.getBytesWritten());
+  delete[] pdxStream;
+}
 
+PdxRemoteWriter PdxHelper::createPdxRemoteWriter(
+    const std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry,
+    const std::shared_ptr<PdxSerializable>& pdxObject, DataOutput& output,
+    const std::string& className) {
+  if (auto pd = pdxTypeRegistry->getPreserveData(pdxObject)) {
```

Comment:
Why does PdxRemoteWriter express a choice between these two constructors? It can perform this check internally on its own. When would you not want to create a writer with a merged type, if it were available? We should review these use cases and probably move this logic to within PdxRemoteWriter. Then we won't need this whole method to begin with.